python/lance/types/box.py

```diff
@@ -30,84 +33,211 @@ def __init__(self):
-            If n = len(self), m = len(others) then returned ious should
-            have shape (n, m)
-        """
-        if not isinstance(others, Box2dArray):
-            raise TypeError("Only box2d knows iou")
-        area_self = self.area()
-        if self is others:
-            area_others = area_self
-        else:
-            area_others = others.area()
-        xmin_inter = np.maximum(self.xmin[:, np.newaxis], others.xmin)
-        ymin_inter = np.maximum(self.ymin[:, np.newaxis], others.ymin)
-        xmax_inter = np.minimum(self.xmax[:, np.newaxis], others.xmax)
-        ymax_inter = np.minimum(self.ymax[:, np.newaxis], others.ymax)
-        intersection = np.maximum(xmax_inter - xmin_inter + 1, 0) * np.maximum(
-            ymax_inter - ymin_inter + 1, 0
+class Box3dType(LanceType):
+    """
+    A bounding box in 3-dimensional space. Each box is represented by
+    2 3-dimensional points
+    """
+
+    def __init__(self):
+        super(Box3dType, self).__init__(
+            pa.list_(pa.float64(), list_size=6),
+            "box3d",
         )
-        union = area_self[:, np.newaxis] + area_others - intersection
-        return intersection / union
 
-    def area(self) -> np.ndarray:
-        """Compute the area of boxes"""
-        return (self.xmax - self.xmin + 1) * (self.ymax - self.ymin + 1)
+    @property
+    def ndims(self):
+        return 3
+
+    def __arrow_ext_class__(self):
+        return Box3dArray
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        return Box3dType()
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+
+class BoxNdArray(pa.ExtensionArray, ABC):
+    @property
+    @abstractmethod
+    def ndims(self):
+        """The number of dimensions of each bounding box in this array"""
+        pass
 
     def flatten(self):
+        """
+        Return a flattened 1-d array
+        """
         # If this is a slice, the default Arrow behavior does not
         # return the sliced values properly
         values = self.storage.values
         # If this is a full-length array then just default Arrow
-        if len(self) == len(values) / 4:
+        if len(self) == len(values) / (self.ndims * 2):
             return values
         else:
             # If this is a slice then we need to compute manually
             return pc.list_flatten(self.storage)
 
     def to_numpy(self, zero_copy_only=True):
+        """
+        Return a homogeneous 2-d numpy array where each row represents the
+        2 points defining each bounding box
+
+        For a 2-d box, each row will have (xmin, ymin, xmax, ymax)
+        For a 3-d box, each row will have (xmin, ymin, zmin, xmax, ymax, zmax)
+        """
         return (
             self.flatten()
             .to_numpy(zero_copy_only=zero_copy_only)
-            .reshape((len(self), 4))
+            .reshape((len(self), (self.ndims * 2)))
         )
 
+    def get_length(self, axis: Union[int, str]):
```

Comment:
Shall we make a new name of it?
i.e, a 2d box (rectangle) has width / length , but this `get_length()` is actually capable to get both IIUC.

Reply:
good point. This is the edge length along a given dimension. 
how about `get_edge_len` ?